pkg/trojan/message.go

```diff
@@ -0,0 +1,283 @@
+	m.Topic = topic
+	m.Payload = payload
+	m.padding = padding
+
+	hashPool := bmtlegacy.NewTreePool(hashFunc, swarm.Branches, bmtlegacy.PoolSize)
+	hasher = bmtlegacy.New(hashPool)
+
+	return *m, nil
+}
+
+// Wrap creates a new trojan chunk for the given targets and Message
+// a trojan chunk is a content-addressed chunk made up of span, a nonce, and a payload which contains the Message
+// the chunk address will have one of the targets as its prefix and thus will be forwarded to the neighbourhood of the recipient overlay address the target is derived from
+func (m *Message) Wrap(targets Targets) (swarm.Chunk, error) {
+	if err := checkTargets(targets); err != nil {
+		return nil, err
+	}
+
+	span := make([]byte, 8)
+	// 4064 bytes for trojan message + 32 bytes for nonce = 4096 bytes as payload for resulting chunk
+	binary.LittleEndian.PutUint64(span, swarm.ChunkSize)
+
+	return m.toChunk(targets, span)
+}
+
+// Unwrap creates a new trojan message from the given chunk payload
+// this function assumes the chunk has been validated as a content-addressed chunk
+// it will return the resulting message if the unwrapping is successful, and an error otherwise
+func Unwrap(c swarm.Chunk) (*Message, error) {
+	d := c.Data()
+
+	// unmarshal chunk payload into message
+	m := new(Message)
+	// first 40 bytes are span + nonce
+	err := m.UnmarshalBinary(d[40:])
+
+	return m, err
+}
+
+// IsPotential returns true if the given chunk is a potential trojan
+func IsPotential(c swarm.Chunk) bool {
+	// chunk must be content-addressed to be trojan
+	if c.Type() != swarm.ContentAddressed {
+		return false
+	}
+
+	data := c.Data()
+	// check for minimum chunk data length
+	// span (8) + nonce (32) + length (2) + topic (32) = 74
+	if len(data) < 74 {
+		return false
+	}
+
+	// check for valid trojan message length in bytes #41 and #42
+	messageLen := int(binary.BigEndian.Uint16(data[40:42]))
+	return 74+messageLen <= len(data)
+}
+
+// checkTargets verifies that the list of given targets is non empty and with elements of matching size
+func checkTargets(targets Targets) error {
+	if len(targets) == 0 {
+		return ErrEmptyTargets
+	}
+	validLen := len(targets[0]) // take first element as allowed length
+	for i := 1; i < len(targets); i++ {
+		if len(targets[i]) != validLen {
+			return ErrVarLenTargets
+		}
+	}
+	return nil
+}
+
+// toChunk finds a nonce so that when the given trojan chunk fields are hashed, the result will fall in the neighbourhood of one of the given targets
+// this is done by iteratively enumerating different nonces until the BMT hash of the serialization of the trojan chunk fields results in a chunk address that has one of the targets as its prefix
+// the function returns a new chunk, with the found matching hash to be used as its address,
+// and its data set to the serialization of the trojan chunk fields which correctly hash into the matching address
+func (m *Message) toChunk(targets Targets, span []byte) (swarm.Chunk, error) {
```

Comment:
I am not so sure about this approach. FIrst of all I am not sure about the use-case where you'd want to send to "any-of-the-above" instead of "all-of-the-above" addresses. Perhaps it's in the book somewhere, of course.

But in any case, doesn't it make more sense to spawn parallell mine operations here? That would reduce the complexity of code in the core message component, pushing the responsibility to a(n optional) "miner driver" above that would orchestrate it, and thus probably ease debugging efforts.

Reply:
agreed on the [parallelization](https://github.com/ethersphere/swarm/issues/2157) point.

regarding "any of the above" vs "all of the above": it will take longer to mine to more than 1 address. 

in terms of global pinning, we just need 1 target. if we are talking about a different use case, such as "message _all_ of these nodes", then yes, we'd need to change this.

maybe this distinction should be implemented. but we'd need to be careful in terms of code structure

Reply:
match a prefix of length d on any of n requires on averages 2^d/n cycles
the use-case is that typically one swarm node will operate multiple overlays each of which can serve as your resident address to receive messages on 'locally'

Reply:
- as @mortelli said "all-of-the-above" does not fall under global pinning 

- parallel mining should be done using a miner as you said.. 
  We would take that as a improvement (see issue created #444 ).